src/Platforms.jl

```diff
@@ -14,3 +16,201 @@ Pkg.BinaryPlatforms.triplet(::AnyPlatform) = "any"
+        "^",
+        c(arch_mapping),
+        c(platform_mapping),
+        c(libc_mapping),
+        c(call_abi_mapping),
+        c(libgfortran_version_mapping),
+        c(libstdcxx_version_mapping),
+        c(cxxstring_abi_mapping),
+        "\$",
+    ))
+
+    m = match(triplet_regex, machine)
+    if m !== nothing
+        # Helper function to find the single named field within the giant regex
+        # that is not `nothing` for each mapping we give it.
+        get_field(m, mapping) = begin
+            for k in keys(mapping)
+                if m[k] !== nothing
+                    strk = string(k)
+                    # Convert our sentinel `nothing` values to actual `nothing`
+                    if endswith(strk, "_nothing")
+                        return nothing
+                    end
+                    # Convert libgfortran/libstdcxx version numbers
+                    if startswith(strk, "libgfortran")
+                        return VersionNumber(parse(Int,strk[12:end]))
+                    elseif startswith(strk, "libstdcxx")
+                        return VersionNumber(3, 4, parse(Int,strk[10:end]))
+                    else
+                        return k
+                    end
+                end
+            end
+        end
+
+        # Extract the information we're interested in:
+        arch = get_field(m, arch_mapping)
+        platform = get_field(m, platform_mapping)
+        libc = get_field(m, libc_mapping)
+        call_abi = get_field(m, call_abi_mapping)
+        libgfortran_version = get_field(m, libgfortran_version_mapping)
+        libstdcxx_version = get_field(m, libstdcxx_version_mapping)
+        cxxstring_abi = get_field(m, cxxstring_abi_mapping)
+
+        # First, figure out what platform we're dealing with, then sub that off
+        # to the appropriate constructor.  If a constructor runs into trouble,
+        # catch the error and return `UnknownPlatform()` here to be nicer to client code.
+        ctors = Dict(:darwin => MacOS, :mingw32 => Windows, :freebsd => FreeBSD, :linux => Linux)
+        try
+            T = ctors[platform]
+            compiler_abi = CompilerABI(;
+                libgfortran_version=libgfortran_version,
+                libstdcxx_version=libstdcxx_version,
+                cxxstring_abi=cxxstring_abi
+            )
+            return T(arch, libc=libc, call_abi=call_abi, compiler_abi=compiler_abi)
+        catch
+        end
+    end
+
+    return nothing
+end
+
+function Base.tryparse(::Type{ExtendedPlatform}, str::AbstractString)
+    split_str = split(str, "-")
+    extra_str = String[]
+    base_platform = platform_key_abi()
+    # Try to parse the longest string possible with the standard parser
+    while length(split_str) > 0
+        base_platform = tryparse_standard_platform(join(split_str, '-'))
+        if base_platform !== nothing
+            break
+        end
+        push!(extra_str, pop!(split_str))
+    end
```

Comment:
I wonder if we can take the split pieces, then just filter out anything that doesn't have a `+` in it, then past those pieces to `tryparse_standard_platform`?  Anything that is "extended" should have `+` in it somewhere, and I don't think `+` can legally appear in a standard platform.